src/main/java/io/vertx/mqtt/impl/MqttServerConnection.java

```diff
@@ -87,7 +90,17 @@ void handleMessage(Object msg) {
 
       DecoderResult result = mqttMessage.decoderResult();
       if (result.isFailure()) {
-        chctx.pipeline().fireExceptionCaught(result.cause());
+        Throwable cause = result.cause();
+        if (cause instanceof MqttUnacceptableProtocolVersionException) {
+
+          MqttMessage connack = MqttEndpointImpl.createConnack(
+            MqttConnectReturnCode.CONNECTION_REFUSED_UNACCEPTABLE_PROTOCOL_VERSION,
+            false);
+          so.writeMessage(connack);
+          so.close();
+        } else {
+          chctx.pipeline().fireExceptionCaught(result.cause());
+        }
```

Comment:
Thinking about this approach it's clear why you did that btw ...
The `connack` method in the `MqttEndpointImpl` already handles closing the connection if the returncode is not about connection accepted so we are duplicating the logic here.
It's also true that here we don't have an `MqttEndpoint` instance which is created later.
Anyway, in order to use the same logic, I would add a new ctor to the `MqttEndpointImpl` accepting just the `conn` parameter and then using the regular `connack` method on it to return the unacceptable protocol version error.
Wdyt?

Reply:
I believe it cannot work because the message we receive is partially decoded as it is a `MqttMessage` instance that is not an  `MqttConnectMessage` instance (because it cannot be constructed with an invalid version protocol version) so we cannot really initialise the endpoint.

Reply:
For this reason, I was talking about a new ctor where the `MqttEndpointImpl` is initialized just with the connection not with "will" information for example that we don't have of course.
Having an endpoint instance helps anyway to avoid code duplication.

Reply:
to be honest this will be an artificial control flow to me, there is not much duplicated here

Reply:
the code is pretty straightforward and is implemented according to the spec when the exception is actually available

Reply:
I reworked the code to do like you said, my personal opinion is stil what I said above.

we need to get this merged soon for 3.8 release

Reply:
I think we had a misunderstanding I meant just:

```java
Throwable cause = result.cause();
if (cause instanceof MqttUnacceptableProtocolVersionException) {
    endpoint = new MqttEndpointImpl(so);
    endpoint.reject(MqttConnectReturnCode.CONNECTION_REFUSED_UNACCEPTABLE_PROTOCOL_VERSION);
}
chctx.pipeline().fireExceptionCaught(result.cause());
return;
```
whitout changing anything else in the handleConnect but adding just a ctor for MqttEndpointImpl with so as parameter.

Reply:
it's not clear what I should do perhaps you can modify yourself the PR and the right commit ?